README.md
[English](https://github.com/bytes7bytes7/problem_details/blob/master/README.md)
| [Русский](https://github.com/bytes7bytes7/problem_details/blob/master/resources/l10n/ru_RU/README.md)

# Problem Details

Implementation of ProblemDetails object (for more info see https://www.rfc-editor.org/rfc/rfc7807)

- [How to use](#how-to-use)
  - [ProblemDetails](#problemdetails)
  - [ProblemDetails.raw](#problemdetailsraw)
  - [Additional fields](#additional-fields)
  - [toJson](#tojson)
  - [fromJson](#fromjson)

## How to use

### ProblemDetails

You can use unnamed factory constructor `ProblemDetails` to create `ProblemDetails`:

```dart
void main() {
  final problemDetails = ProblemDetails();
}
```

`status` is `500` by default.

`ProblemDetails` constructor fulfills fields: `type` and `title` based on `status`. But, it only
works for a limited set of `status` values. The full set of `status` values you can see
[here](https://github.com/bytes7bytes7/problem_details/blob/master/lib/src/problem_details_defaults.dart)
.

For example:

```dart
void main() {
  final problemDetails = ProblemDetails(status: 403);

  // prints: https://www.rfc-editor.org/rfc/rfc7231#section-6.5.3
  print(problemDetails.type);

  // prints: Forbidden
  print(problemDetails.title);
}
```

If `ProblemDetails` does not have default values of `type`, `title` for a certain `status`,
then `type` is set to `about:blank`.

```dart
void main() {
  final problemDetails = ProblemDetails(status: 1);

  // prints: about:blank
  print(problemDetails.type);
}
```

### ProblemDetails.raw

Use `ProblemDetails.raw` if you want to set all values manually:

```dart
void main() {
  final problemDetails = ProblemDetails.raw(status: 403);

  // prints: null
  print(problemDetails.type);

  // prints: null
  print(problemDetails.title);
}
```

### Additional fields

If you want to add new fields to `ProblemDetails`, use `extensions`:

```dart
void extensions() {
  var problemDetails = ProblemDetails(
    extensions: {
      'key1': 'value',
      'key2': null,
      'key3': {
        'key4': [1, 2, 3],
      },
    },
  );

  // prints:
  //  {
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    }
  //  }
  print(problemDetails.extensions);
}
```

You can change/add a field by using `addOrUpdateExtension` method:

```dart
void main() {
  var problemDetails = ProblemDetails(
    extensions: {
      'key1': 'value',
      'key2': null,
      'key3': {
        'key4': [1, 2, 3],
      },
    },
  );

  problemDetails = problemDetails.addOrUpdateExtension('KEY', 'VALUE');

  // prints:
  //  {
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    },
  //    'KEY': 'VALUE'
  //  }
  print(problemDetails.extensions);
}
```

You can change/add multiple fields by using `addOrUpdateExtensions` method:

```dart
void main() {
  var problemDetails = ProblemDetails(
    extensions: {
      'key1': 'value',
      'key2': null,
      'key3': {
        'key4': [1, 2, 3],
      },
    },
  );

  problemDetails = problemDetails.addOrUpdateExtensions(
    {
      'a': 1,
      'b': 2,
    },
  );

  // prints:
  //  {
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    },
  //    'a': 1,
  //    'b': 2
  //  }
  print(problemDetails.extensions);
}
```

You can remove a field by using `removeExtension` method:

```dart
void main() {
  var problemDetails = ProblemDetails(
    extensions: {
      'key1': 'value',
      'key2': null,
      'key3': {
        'key4': [1, 2, 3],
      },
    },
  );

  problemDetails = problemDetails.removeExtension('key1');

  // prints:
  //  {
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    }
  //  }
  print(problemDetails.extensions);
}
```

### toJson

`toJson` method converts `ProblemDetails` into `Map`.

> **_NOTE:_** Standard fields (`status`, `type`,
> `title`, `details`, `instance`) are not included in a Map, if a field is null.

> **_IMPORTANT:_** Content of `extensions` is put on the top level of a Map.

```dart
void main() {
  final problemDetails = ProblemDetails(
    status: 403,
    extensions: {
      'key1': 'value',
      'key2': null,
      'key3': {
        'key4': [1, 2, 3],
      },
    },
  );

  final jsonData = problemDetails.toJson();

  // prints:
  //  {
  //    status: 403,
  //    type: https://www.rfc-editor.org/rfc/rfc7231#section-6.5.3,
  //    title: Forbidden,
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    }
  //  }
  print(jsonData);
}
```

### fromJson

`fromJson` constructor parses `Map` into `ProblemDetails`.

> **_IMPORTANT:_** Non-standard fields of `ProblemDetails`
> (i. e. all fields except `status`, `type`,
> `title`, `details`, `instance`) convert in `extensions`.

```dart
void main() {
  var problemDetails = ProblemDetails(
    status: 403,
    extensions: {
      'key1': 'value',
      'key2': null,
      'key3': {
        'key4': [1, 2, 3],
      },
    },
  );

  final jsonData = problemDetails.toJson();

  problemDetails = ProblemDetails.fromJson(jsonData);

  // prints:
  //  ProblemDetails {
  //    status: 403,
  //    type: https://www.rfc-editor.org/rfc/rfc7231#section-6.5.3,
  //    title: Forbidden,
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    }
  //  }
  print(problemDetails);
}
```


CHANGELOG.md
## 0.0.2

- Remove `dart:io`

## 0.0.1

- Add `ProblemDetails` class.
- Add `ProblemDetails`, `ProblemDetails.raw` constructors.
- Add `fromJson` and `toJson` methods in `ProblemDetails`.
- Override `toString` in `ProblemDetails`.
- Add defaults for `ProblemDetails`.
- Add `addOrUpdateExtension`, `addOrUpdateExtensions`, `removeExtension` methods for
  modifying `extensions` of `ProblemDetails`.


test/problem_details_test.dart
import 'package:problem_details/problem_details.dart';
import 'package:test/test.dart';

void main() {
  group('ProblemDetails constructor', () {
    test('status is 500 by default', () {
      final problemDetails = ProblemDetails();

      expect(problemDetails.status, 500);
    });

    test('sets default `type` if it is not specified', () {
      final problemDetails = ProblemDetails();

      expect(problemDetails.type, isNotNull);
    });

    test('sets default `title` if it is not specified', () {
      final problemDetails = ProblemDetails();

      expect(problemDetails.title, isNotNull);
    });

    test(
        'sets `type` to `about:blank` if it is not specified '
        'and there is not default value', () {
      // Sets status to non-existent value, to make sure that
      // the default value for `type` can not be found.
      final problemDetails = ProblemDetails(status: 1);

      expect(problemDetails.type, 'about:blank');
    });
  });

  group('ProblemDetails toJson()', () {
    test('removes pairs where a value is null (only for "standard" fields)',
        () {
      final problemDetails = ProblemDetails.raw(
        status: 1,
      );

      expect(problemDetails.toJson(), {'status': 1});
    });

    test('puts data of `extensions` on the top level of Map', () {
      final extensions = {
        'key1': 'value1',
        'key2': {
          'key3': 'value3',
        },
      };
      final problemDetails = ProblemDetails.raw(
        status: 1,
        extensions: extensions,
      );

      expect(
        problemDetails.toJson(),
        <String, Object?>{
          'status': 1,
        }..addAll(extensions),
      );
    });
  });

  group('fromJson()', () {
    test('all data except "standard" fields are related to `extensions`', () {
      final extensions = {
        'key1': 'value1',
        'key2': {
          'key3': 'value3',
        },
      };
      final problemDetails = ProblemDetails.raw(
        status: 1,
        extensions: extensions,
      );
      final jsonData = problemDetails.toJson();

      expect(
        ProblemDetails.fromJson(jsonData).extensions,
        extensions,
      );
    });
  });

  group('addOrUpdateExtension', () {
    test('adds new pair to `extensions` if it is not exist', () {
      final key = 'key';
      final value = 'value';
      final problemDetails = ProblemDetails();

      expect(
        problemDetails.addOrUpdateExtension(key, value).extensions,
        {key: value},
      );
    });

    test('updates a pair to `extensions` if exists', () {
      final key = 'key';
      final value1 = 'value1';
      final value2 = 'value2';
      final problemDetails = ProblemDetails(
        extensions: {
          key: value1,
        },
      );

      expect(
        problemDetails.addOrUpdateExtension(key, value2).extensions,
        {key: value2},
      );
    });
  });

  group('addOrUpdateExtensions', () {
    test('adds new pairs to `extensions` if they are not exist', () {
      final newExtensions = {
        'key1': 'value1',
        'key2': 'value2',
      };
      final problemDetails = ProblemDetails();

      expect(
        problemDetails.addOrUpdateExtensions(newExtensions).extensions,
        newExtensions,
      );
    });

    test('updates pairs to `extensions` if they exist', () {
      final key1 = 'key1';
      final key2 = 'key2';
      final oldExtensions = {
        key1: 'value1',
        key2: 'value2',
      };
      final newExtensions = {
        key1: 'value3',
        key2: 'value4',
      };
      final problemDetails = ProblemDetails(
        extensions: oldExtensions,
      );

      expect(
        problemDetails.addOrUpdateExtensions(newExtensions).extensions,
        newExtensions,
      );
    });
  });

  group('removeExtension', () {
    test('removes a pair from `extensions` if it exists', () {
      final key = 'key';
      final problemDetails = ProblemDetails(
        extensions: {
          key: 'value',
        },
      );

      expect(
        problemDetails.removeExtension(key).extensions,
        <String, Object?>{},
      );
    });
  });
}


lib/src/problem_details.dart
part 'problem_details_defaults.dart';

const _statusKey = 'status';
const _typeKey = 'type';
const _titleKey = 'title';
const _detailKey = 'detail';
const _instanceKey = 'instance';

/// Machine-readable details of errors for a HTTP response.
///
/// For more info see https://www.rfc-editor.org/rfc/rfc7807
class ProblemDetails {
  /// The HTTP status code generated by the origin server for
  /// this occurrence of the problem.
  final int status;

  /// A URI reference that identifies the problem type.
  /// This specification encourages that, when dereferenced, it provide
  /// human-readable documentation for the problem type.
  final String? type;

  /// A short, human-readable summary of the problem type.
  final String? title;

  /// A human-readable explanation specific to this occurrence
  /// of the problem.
  final String? detail;

  /// A URI reference that identifies the specific occurrence
  /// of the problem.
  final String? instance;

  /// Additional information about problem.
  final Map<String, Object?> extensions;

  factory ProblemDetails({
    int? status,
    String? type,
    String? title,
    String? detail,
    String? instance,
    Map<String, Object?>? extensions,
  }) {
    status ??= 500;

    final defaults = _problemDetailsDefaults[status];

    if (defaults != null) {
      type ??= defaults.link;
      title ??= defaults.title;
    }

    type ??= 'about:blank';

    return ProblemDetails.raw(
      status: status,
      type: type,
      title: title,
      detail: detail,
      instance: instance,
      extensions: extensions,
    );
  }

  const ProblemDetails.raw({
    required this.status,
    this.type,
    this.title,
    this.detail,
    this.instance,
    Map<String, Object?>? extensions,
  }) : extensions = extensions ?? const {};

  factory ProblemDetails.fromJson(Map<String, Object?> json) {
    final status = json.remove(_statusKey) as int;
    final type = json.remove(_typeKey) as String?;
    final title = json.remove(_titleKey) as String?;
    final detail = json.remove(_detailKey) as String?;
    final instance = json.remove(_instanceKey) as String?;

    return ProblemDetails.raw(
      status: status,
      type: type,
      title: title,
      detail: detail,
      instance: instance,
      extensions: json,
    );
  }

  @override
  String toString() => 'ProblemDetails ${toJson()}';

  Map<String, Object?> toJson() {
    final result = <String, Object?>{
      _statusKey: status,
    };

    if (type != null) {
      result[_typeKey] = type;
    }

    if (title != null) {
      result[_titleKey] = title;
    }

    if (detail != null) {
      result[_detailKey] = detail;
    }

    if (instance != null) {
      result[_instanceKey] = instance;
    }

    result.addAll(extensions);

    return result;
  }

  ProblemDetails addOrUpdateExtension(String key, Object? value) {
    return ProblemDetails.raw(
      status: status,
      type: type,
      title: title,
      detail: detail,
      instance: instance,
      extensions: Map.from(extensions)..[key] = value,
    );
  }

  ProblemDetails addOrUpdateExtensions(Map<String, Object?> data) {
    return ProblemDetails.raw(
      status: status,
      type: type,
      title: title,
      detail: detail,
      instance: instance,
      extensions: Map.from(extensions)..addAll(data),
    );
  }

  ProblemDetails removeExtension(String key) {
    return ProblemDetails.raw(
      status: status,
      type: type,
      title: title,
      detail: detail,
      instance: instance,
      extensions: Map.from(extensions)..remove(key),
    );
  }
}


lib/src/problem_details_defaults.dart
part of 'problem_details.dart';

class _DefaultsValue {
  const _DefaultsValue({
    required this.link,
    required this.title,
  });

  final String link;
  final String title;
}

/// Predefined info for Problem Details.
const _problemDetailsDefaults = <int, _DefaultsValue>{
  400: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.1',
    title: 'Bad Request',
  ),
  401: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7235#section-3.1',
    title: 'Unauthorized',
  ),
  403: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.3',
    title: 'Forbidden',
  ),
  404: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.4',
    title: 'Not Found',
  ),
  405: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.5',
    title: 'Method Not Allowed',
  ),
  406: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.6',
    title: 'Not Acceptable',
  ),
  407: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7235#section-3.2',
    title: 'Proxy Authentication Required',
  ),
  408: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.7',
    title: 'Request Timeout',
  ),
  409: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.8',
    title: 'Conflict',
  ),
  410: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.9',
    title: 'Gone',
  ),
  411: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.10',
    title: 'Length Required',
  ),
  413: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.11',
    title: 'Payload Too Large',
  ),
  414: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.12',
    title: 'URI Too Long',
  ),
  415: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.13',
    title: 'Unsupported Media Type',
  ),
  417: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.14',
    title: 'Expectation Failed',
  ),
  422: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc4918#section-11.2',
    title: 'Unprocessable Entity',
  ),
  426: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.5.15',
    title: 'Upgrade Required',
  ),
  500: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.6.1',
    title: 'Internal Server Error',
  ),
  501: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.6.2',
    title: 'Not Implemented',
  ),
  502: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.6.3',
    title: 'Bad Gateway',
  ),
  503: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.6.4',
    title: 'Service Unavailable',
  ),
  504: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.6.5',
    title: 'Gateway Timeout',
  ),
  505: _DefaultsValue(
    link: 'https://www.rfc-editor.org/rfc/rfc7231#section-6.6.6',
    title: 'HTTP Version Not Supported',
  ),
};


example/problem_details_example.dart
// ignore_for_file: avoid_print

import 'package:problem_details/problem_details.dart';

void main() {
  predefinedWithNoParams();
  predefined();
  notPredefined();
  raw();
  extensions();
  toJson();
  fromJson();
}

void predefinedWithNoParams() {
  /// If you do not specify `status`, then it sets to 500 by default.
  final problemDetails = ProblemDetails();

  /// prints: 500
  print(problemDetails.status);
}

void predefined() {
  /// ProblemDetails has some predefined `type`s and `title`s for
  /// some status codes.
  /// For example, for 403 there is predefined `type` and `title`.
  /// Thus, ProblemDetails fulfills `type` and `title` with defaults.
  final problemDetails = ProblemDetails(status: 403);

  // prints: https://www.rfc-editor.org/rfc/rfc7231#section-6.5.3
  print(problemDetails.type);

  // prints: Forbidden
  print(problemDetails.title);
}

void notPredefined() {
  /// If ProblemDetails does not have predefined `type` and `title` for
  /// specified status, then `type` is set to 'about:blank'
  final problemDetails = ProblemDetails(status: 1);

  // prints: about:blank
  print(problemDetails.type);
}

void raw() {
  /// If you want to set all fields manually without autocomplete
  /// with defaults, use `ProblemDetails.raw` constructor.
  final problemDetails = ProblemDetails.raw(status: 403);

  // prints: null
  print(problemDetails.type);

  // prints: null
  print(problemDetails.title);
}

void extensions() {
  /// If you want to add custom fields to ProblemDetails, use `extensions`.
  var problemDetails = ProblemDetails(
    extensions: {
      'key1': 'value',
      'key2': null,
      'key3': {
        'key4': [1, 2, 3],
      },
    },
  );

  // prints:
  //  {
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    }
  //  }
  print(problemDetails.extensions);

  /// You can modify single pair of `extensions` with `addOrUpdateExtension`.
  problemDetails = problemDetails.addOrUpdateExtension('KEY', 'VALUE');

  // prints:
  //  {
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    },
  //    'KEY': 'VALUE'
  //  }
  print(problemDetails.extensions);

  /// You can modify multiple pairs of `extensions` with
  /// `addOrUpdateExtensions`.
  problemDetails = problemDetails.addOrUpdateExtensions(
    {
      'a': 1,
      'b': 2,
    },
  );

  // prints:
  //  {
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    },
  //    'KEY': 'VALUE',
  //    'a': 1,
  //    'b': 2
  //  }
  print(problemDetails.extensions);

  /// You can remove some pair of `extensions` with `removeExtension`.
  problemDetails = problemDetails.removeExtension('KEY');

  // prints:
  //  {
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    },
  //    'a': 1,
  //    'b': 2
  //  }
  print(problemDetails.extensions);
}

void toJson() {
  final problemDetails = ProblemDetails(
    status: 403,
    extensions: {
      'key1': 'value',
      'key2': null,
      'key3': {
        'key4': [1, 2, 3],
      },
    },
  );

  /// `toJson` converts `ProblemDetails` to a Map.
  /// Standard fields (`status`, `type`, `title`, `details`, `instance`)
  /// are not included in a Map if a field is null.
  /// Content of `extensions` field is put on the top level of a Map.
  final jsonData = problemDetails.toJson();

  // prints:
  //  {
  //    status: 403,
  //    type: https://www.rfc-editor.org/rfc/rfc7231#section-6.5.3,
  //    title: Forbidden,
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    }
  //  }
  print(jsonData);
}

void fromJson() {
  var problemDetails = ProblemDetails(
    status: 403,
    extensions: {
      'key1': 'value',
      'key2': null,
      'key3': {
        'key4': [1, 2, 3],
      },
    },
  );

  final jsonData = problemDetails.toJson();

  /// `ProblemDetails.fromJson` constructor parse Map to ProblemDetails.
  problemDetails = ProblemDetails.fromJson(jsonData);

  // prints:
  //  ProblemDetails {
  //    status: 403,
  //    type: https://www.rfc-editor.org/rfc/rfc7231#section-6.5.3,
  //    title: Forbidden,
  //    key1: value,
  //    key2: null,
  //    key3: {
  //      key4: [1, 2, 3]
  //    }
  //  }
  print(problemDetails);
}
